src/main/kotlin/de/smartsquare/starter/mqtt/MqttPublisher.kt

```diff
@@ -37,17 +43,97 @@ class Mqtt5Publisher(private val adapter: MqttMessageAdapter, client: Mqtt5Clien
 
     private val asyncClient = client.toAsync()
 
+    /**
+     * Options for publishing a message.
+     */
+    data class PublishingOptions(
+        val retain: Boolean = false,
+        val messageExpiryInterval: Long = MqttPublish.NO_MESSAGE_EXPIRY,
+        val payloadFormatIndicator: Mqtt5PayloadFormatIndicator? = null,
+        val contentType: String? = null,
+        val responseTopic: MqttTopic? = null,
+        val correlationData: ByteArray? = null,
+        val userProperties: Mqtt5UserProperties = Mqtt5UserProperties.of(),
+    ) {
+        companion object {
+            /**
+             * Creates a new [Builder] for [PublishingOptions]. With the given [original] as a base.
+             * If no base is given, the default values from [PublishingOptions] are used.
+             */
+            @JvmStatic
+            @JvmOverloads
+            fun builder(original: PublishingOptions = PublishingOptions()) = object : Builder {
+                private var state = original
+
+                override fun retain(retain: Boolean): Builder = apply {
+                    state = state.copy(retain = retain)
+                }
+
+                override fun messageExpiryInterval(messageExpiryInterval: Long): Builder = apply {
+                    state = state.copy(messageExpiryInterval = messageExpiryInterval)
+                }
+
+                override fun payloadFormatIndicator(payloadFormatIndicator: Mqtt5PayloadFormatIndicator?): Builder =
+                    apply { state = state.copy(payloadFormatIndicator = payloadFormatIndicator) }
+
+                override fun contentType(contentType: String?): Builder = apply {
+                    state = state.copy(contentType = contentType)
+                }
+
+                override fun responseTopic(responseTopic: MqttTopic?): Builder = apply {
+                    state = state.copy(responseTopic = responseTopic)
+                }
+
+                override fun correlationData(correlationData: ByteArray?): Builder = apply {
+                    state = state.copy(correlationData = correlationData)
+                }
+
+                override fun userProperties(userProperties: Mqtt5UserProperties): Builder = apply {
+                    state = state.copy(userProperties = userProperties)
+                }
+
+                override fun build(): PublishingOptions = state
+            }
+        }
+
+        /**
+         * Fluent builder for [PublishingOptions]. It is initialized with the default values from [PublishingOptions].
+         */
+        interface Builder {
```

Comment:
Could this not be a `class` with the implementation from above? I don't see any reason to have abstraction here.

Reply:
Done